node-wasm/src/worker/channel.rs

```diff
@@ -0,0 +1,191 @@
+use std::fmt::{self, Debug};
+
+use serde_wasm_bindgen::{from_value, to_value};
+use tokio::sync::{mpsc, Mutex};
+use tracing::{debug, error, info, warn};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::spawn_local;
+use web_sys::{MessageEvent, MessagePort, SharedWorker};
+
+use crate::utils::WorkerSelf;
+use crate::worker::commands::{NodeCommand, WorkerResponse};
+use crate::worker::WorkerError;
+
+type WireMessage = Option<Result<WorkerResponse, WorkerError>>;
+type WorkerClientConnection = (MessagePort, Closure<dyn Fn(MessageEvent)>);
+
+// TODO: cleanup JS objects on drop
+// impl Drop
+pub(crate) struct WorkerClient {
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+    channel: MessagePort,
+    response_channel: Mutex<mpsc::Receiver<WireMessage>>,
+}
+
+impl WorkerClient {
+    pub fn new(channel: MessagePort) -> Self {
+        let (response_tx, response_rx) = mpsc::channel(64);
+
+        let near_tx = response_tx.clone();
+        let onmessage_callback = move |ev: MessageEvent| {
+            let local_tx = near_tx.clone();
+            spawn_local(async move {
+                let message_data = ev.data();
+
+                let data = from_value(message_data)
+                    .map_err(|e| {
+                        error!("could not convert from JsValue: {e}");
+                    })
+                    .ok();
```

Comment:
I believe instead of `None` here, we should generate something like `WorkerError::InvalidResponse` and `WireMessage` be `Result<WorkerResponse, WorkerError>`.

Reply:
Good idea, reworked `WireMessage` to carry a `Result` with a `WorkerError::CouldNotDeserialiseResponse`.